feat(cli): warn if incompatible node/package versions detected

danielroe opened the pull request:
## Types of changes
- [x] New feature (a non-breaking change which adds functionality)

## Description
closes #8783

## Checklist:
- [x] I have added tests to cover my changes
- [x] All new and existing tests are passing

danielroe committed:
feat(cli): warn if incompatible node/package versions detected

pi0 commented on packages/cli/src/utils/dependencies.js:
```diff
@@ -0,0 +1,49 @@
+
+export function checkDependencies() {
+  const packagesToTrack = ['postcss', 'webpack']
+  const webpackDeps = getPKG('@nuxt/webpack').dependencies
```
An inline list would be more than enough since exceptions are manual anyway (and the fact that some deps are optional like sass-loader)

pi0 replied:
I would use an object syntax like `{ 'sass-loader':  { range: '^10.1.1' }, 'node': { range: '>12', message: 'you are using unsupported NodeJS version' }`

danielroe replied:
Node.js version checking has to be done differently so I think we don't need a single object.

pi0 replied:
Actually it was idea of #8783 for generic version checking system. And we might make smarter warnings by checking if package is explicitly installed by `{rootDir}/package.json` or hoisted. But seems okay for now we might do it later on as a nuxt-contrib project :)

pi0 commented on packages/cli/src/commands/generate.js:
```diff
@@ -61,6 +62,7 @@ export default {
     }
   },
   async run (cmd) {
+    checkDependencies()
```
We shall run check in setup.js instead of commands

pi0 commented on packages/cli/src/utils/dependencies.js:
```diff
@@ -0,0 +1,49 @@
+  if (satisfies(process.version, '^10.0.0')) {
+    consola.warn('Node.js 10 is deprecated and will shortly not be supported.')
+  } else if (!satisfies(process.version, '>12.0.0')) {
+    consola.error('Please use a supported version of Node.js (12+).')
```
IMO thin can be still a warning for You are using an incompatible NodeJS version. An error is not necessarily happening by using node 10

pi0 replied:
Also no exception will be needed anymore for 10.x (version constraints are always updating)

danielroe replied:
I've moved the <10 to a warn (as 10 is). Not sure what you mean by last comment about 10.x.

pi0 replied:
I mean when `>12.0.0` constraint did not met warning can be generic like _You are using an unsupported version of Node.js (x.y.z) and nuxt is likely to be broken. It is recommended to use latest LTS version (https://nodejs.org)_

note that node@10 is only supported by engine. With 2.15 release notes we are also asking to use 12+. And this might be changed in the future too depending on what happens to nuxt dependencies. So there instead of exception for 10.x, we can have a generic advice and single semver constraint.

danielroe replied:
Very helpful - yes, much better approach. I've made this more generic but happy to refactor further either now or in a further PR.

danielroe committed:
refactor: check dependencies in setup

danielroe committed:
refactor: replace dynamic versions with static list

danielroe committed:
test: fix test for node version

danielroe committed:
refactor: make dependency checking slightly more generic

danielroe committed:
Merge branch 'dev' into 8783-incompatible-versions-warning

pi0 commented on packages/cli/src/utils/dependencies.js:
```diff
@@ -0,0 +1,50 @@
+      node: {
+        range: '>12.0.0',
+        message:
+          'You are using an unsupported version of Node.js (x.y.z) and Nuxt is likely to be broken. It is recommended to use the latest LTS version (https://nodejs.org)'
```
👀 👀  🗡️

pi0 replied:
@danielroe Still missing something 😆 (x.y.z). I need to do some refactors before merge will fix it too

pi0 replied:
Didn't noticed replace in bottom :)

danielroe commented on packages/cli/src/utils/dependencies.js:
```diff
@@ -0,0 +1,50 @@
+    },
+    engines: {
+      node: {
+        range: '>12.0.0',
```
```suggestion
        range: '>=12.0.0',
```

danielroe committed:
fix: don't error if node version == `v12.0.0`

danielroe commented on packages/cli/src/utils/dependencies.js:
```diff
@@ -0,0 +1,50 @@
+      node: {
+        range: '>=12.0.0',
+        message:
+          'You are using an unsupported version of Node.js (x.y.z) and Nuxt is likely to be broken. It is recommended to use the latest LTS version (https://nodejs.org)'
```
Add link to LTS release schedule.
```suggestion
          'You are using an unsupported version of Node.js (x.y.z) and Nuxt is likely to be broken. It is recommended to use the latest LTS version (https://nodejs.org/en/about/releases/)'
```

danielroe committed:
fix: add specific link to release schedule

pi0 committed:
chore: always check dependencies

production is probably even more important since servers might use old node version

pi0 committed:
chore: simplify dependencies.js

pi0 committed:
update test

pi0 committed:
Merge branch 'dev' into 8783-incompatible-versions-warning

danielroe commented on packages/cli/test/unit/dependencies.test.js:
```diff
@@ -0,0 +1,25 @@
+        /webpack@.+ is installed but .+ is expected/
+      )
+    )
+    expect(consola.warn).toHaveBeenCalledTimes(1)
```
I think this can be dropped @pi0
